services/notifier/notifier.go

```diff
@@ -0,0 +1,496 @@
+func (n *Notifier) claim(ctx context.Context, workerID string) (*model.Job, model.JobMetadata, error) {
+	claimStartTime := n.now()
+
+	claimedJob, metadata, err := n.repo.Claim(ctx, workerID)
+	if err != nil {
+		n.stats.claimFailedTime.Since(claimStartTime)
+		n.stats.claimFailed.Increment()
+
+		return nil, nil, fmt.Errorf("claiming job: %w", err)
+	}
+
+	n.stats.claimLag.SendTiming(n.now().Sub(claimedJob.CreatedAt))
+	n.stats.claimSucceededTime.Since(claimStartTime)
+	n.stats.claimSucceeded.Increment()
+
+	return claimedJob, metadata, nil
+}
+
+// UpdateClaim updates the notifier with the claimResponse
+// In case if we are not able to update the claim, we are just logging it,
+// maintenance workers can again mark the status as waiting, and it will be again claimed by somebody else.
+// Although, there is a case that it is being picked up, but never getting updated. We can monitor it using claim lag.
+// claim lag also helps us to make sure that even the maintenance workers are able to monitor the jobs correctly.
+func (n *Notifier) UpdateClaim(ctx context.Context, claimedJob *model.ClaimJob, response *model.ClaimJobResponse) {
+	if response.Err != nil {
+		if err := n.repo.OnClaimFailed(ctx, claimedJob.Job, response.Err, n.config.maxAttempt); err != nil {
+			n.stats.claimUpdateFailed.Increment()
+			n.logger.Errorf("update claimed: on claimed failed: %v", err)
+		}
+
+		if claimedJob.Job.Attempt > n.config.maxAttempt {
+			n.stats.abortedRecords.Increment()
+		}
+		return
+	}
+
+	if err := n.repo.OnClaimSuccess(ctx, claimedJob.Job, response.Payload); err != nil {
+		n.stats.claimUpdateFailed.Increment()
+		n.logger.Errorf("update claimed: on claimed success: %v", err)
+	}
+}
+
+// RunMaintenanceWorker re-triggers zombie jobs which were left behind by dead workers in executing state
+// Since it's a blocking call, it should be run in a separate goroutine
+func (n *Notifier) RunMaintenanceWorker(ctx context.Context) error {
+	maintenanceWorkerLockID := murmur3.Sum64([]byte(queueName))
+	maintenanceWorkerLock, err := pglock.NewLock(ctx, int64(maintenanceWorkerLockID), n.db.DB)
+	if err != nil {
+		return fmt.Errorf("creating maintenance worker lock: %w", err)
+	}
+
+	var locked bool
+	defer func() {
+		if locked {
+			if err := maintenanceWorkerLock.Unlock(ctx); err != nil {
+				n.logger.Warnf("unlocking maintenance worker lock: %v", err)
+			}
+		}
+	}()
+
+	for {
+		if locked, err = maintenanceWorkerLock.Lock(ctx); err != nil {
+			n.logger.Warnf("acquiring maintenance worker lock: %v", err)
+		} else if locked {
+			break
+		}
+
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(n.config.jobOrphanTimeout / 5):
+		}
+	}
+
+	for {
+		orphanJobIDs, err := n.repo.OrphanJobIDs(ctx, int(n.config.jobOrphanTimeout/time.Second))
+		if err != nil {
+			if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+				return nil
+			}
+			if pqErr, ok := err.(*pq.Error); ok && pqErr.Code == "57014" {
```

Comment:
This doesn't take into consideration that the error could have been wrapped and yet be a `*pq.Error` under the hood.

```go
package main

import (
	"errors"
	"fmt"

	"github.com/lib/pq"
)

func main() {
	originalErr := &pq.Error{
		Code:             "123",
		Message:          "Hello World",
	}
	wrappedErr := fmt.Errorf("wrapped: %w", originalErr)

	var pqErr *pq.Error
	if errors.As(wrappedErr, &pqErr) && pqErr.Code == "123" {
		fmt.Println("errors.As: pq error code is 123")
	}

	if pqErr, ok := wrappedErr.(*pq.Error); ok && pqErr.Code == "57014" {
		fmt.Println("typeAssertion: pq error code is 123")
	}
}
```

https://go.dev/play/p/kxL4eZG475K?v=

As you can see `typeAssertion: pq error code is 123` did not get printed, but `errors.As: pq error code is 123` did.